209: Add comment fields for translators

Mte90 opened the pull request:
![screenshot_20170826_225814](https://user-images.githubusercontent.com/403283/29745056-147424fa-8ab2-11e7-9e65-bf740e7b12ed.png)

This code save the feedback when the reviewer click to change the status.
The field is disabled if is not the user that written.

I have to improve the check for permissions but actually my implementation is very simple, a new table and is possible to leave only a feedback for translation so I am not sure if it is better to leave a comment with reply or block to changes after the submit.

Ref: https://github.com/GlotPress/GlotPress-WP/issues/209

toolstack commented:
There are a few things I noticed:

1. The text area seems misaligned in the screenshot above.
1. You need to update the db version in glotpress.php otherwise the new table will not be created.
1. There should be a separate button to save the feedback, you shouldn't have to save a translation or a/r/f one to add a note.
1. It might make sense to insert the "by admin @" line inside the note itself when it's saved, otherwise you'll lose any previous time stamps.

It's also throwing two errors with an empty table: 

1. Notice: Undefined offset: 0 in C:\wamp64\www\wp\wp-content\plugins\glotpress\gp-templates\helper-functions.php on line 239
2.  Notice: Trying to get property of non-object in C:\wamp64\www\wp\wp-content\plugins\glotpress\gp-templates\helper-functions.php on line 226

toolstack commented:
One other item, I'm not sure I like "feedbacks", perhaps "notes" would be better?  Or something else?

Mte90 commented:
Yes of course there few issues but was the first draft but I can fix them asap.
I was thinking that was more helpful to save the feedback when there is the approval to avoid another step.

toolstack commented:
It's fine to save it during approval, but there should be a button to save it outside of that process as well, at least in my opinion.

I can imagine an approver going back and forth with the translator a few times before getting it just right and have to "reject" it each time seems a little harsh 😉

Mte90 commented:
Good point!
I will work on that asap on all the feedbacks

toolstack commented:
Something else to think about is creating a "thing" for the "feedbacks" (maybe rename "comments"?) like there is for the translations, projects, etc. (take a look in the gp-includes/things directory).

Mte90 commented:
![screenshot_20170828_225556](https://user-images.githubusercontent.com/403283/29793495-52ff8b86-8c44-11e7-84df-46bf851d413d.png)

Done everything, actually wehn you add a new comment at the end of the text in the textarea automatically add the username and the date and not create duplicate in the database.

Actually is missing only the check of the permission on saving.

Mte90 commented:
I don't like the actual workflow, go to the textarea, scroll bottom, add your text and submit and automatically there is the feedback with appended username and date. The comment will be the same entry in the db with only updated the date and the user author of the last edit.

In any case is working and probably will not to be integrated with #702

toolstack commented:
I'd invert the flow, put new entries at the top.

Perhaps break it up in to two sections, a read only textbox to show the stored notes and a second to write the new addition?

toolstack commented:
Or perhaps when you click "Add feedback" it gives you a dialog box to enter the new feedback in to?

Mte90 commented:
Uhm yes maybe show the old as not editable text outside the textarea and when saved that content is merged

toolstack commented:
Perhaps only let admins edit the old text, just in case they need to remove inappropriate content.

Mte90 commented:
It's not clear to me how works the permission system in glotpress so actually I didn't implemented this part to work on the rest of stuff :-)

toolstack commented:
Permissions are supported through the `GP::$permission` object (see [things/permissions.php](https://github.com/GlotPress/GlotPress-WP/blob/develop/gp-includes/things/permission.php) for more detail.

Basically you can use `user_can()` and `current_user_can()` to test if a user has a specified permission to a specific object type.  The permissions are store in the `gp_permissions` table.

So to edit existing feedback you might do something like:

`GP::$permission->current_user_can( 'admin', 'feedbacks', $project_id )`

But to simply add feedback you could do something like:

`GP::$permission->current_user_can( 'approve', 'translation', $this->id, array( 'translation' => $this ) )`

Actions include: read, write, approve, admin, delete.  But are not limited to those, any action that can be stored in the permissions table can be referenced by the permission functions.

Likewise for object types.

Mte90 commented:
![screenshot_20170829_145728](https://user-images.githubusercontent.com/403283/29822590-4d08fc98-8ccc-11e7-9567-f485239d8fa9.png)

Ok new code with few bugfix now I will work on the permission part :-)

toolstack commented on glotpress.php:
```diff
@@ -27,7 +27,7 @@
 
 define( 'GP_VERSION', '2.4.0-alpha' );
-define( 'GP_DB_VERSION', '980' );
+define( 'GP_DB_VERSION', '981' );
```
I should have mentioned, we usually bump the number in increments of 10, don't ask me why, just the way it's been done in the past.

Mte90 replied:
fixed

Mte90 commented:
![screenshot_20170829_153344](https://user-images.githubusercontent.com/403283/29823619-7cb88c9e-8ccf-11e7-936b-a6f21ec64ede.png)
Permissions implemented, the button label change by permission and enabled to submit both the text if admin.

toolstack commented on gp-templates/helper-functions.php:
```diff
@@ -172,44 +172,91 @@ function textareas( $entry, $permissions, $index = 0 ) {
+		$button_label = __( 'Add feedback', 'glotpress' );
+	?>
+	<?php
+		if ( GP::$permission->current_user_can( 'admin', 'feedbacks', $project_id ) ) {
```
`$project_id` isn't defined here you have to look up the project first.

Mte90 replied:
fixed

toolstack commented on gp-includes/things/translation.php:
```diff
@@ -628,13 +628,50 @@ public function set_status( $status ) {
+		if ( ! GP::$permission->current_user_can( 'approve', 'translation', $entry->id, array( 'translation' => $entry ) ) ) {
+			return false;
+		}
+		if ( GP::$permission->current_user_can( 'admin', 'feedbacks', $project_id ) ) {
```
Same as above, $project_id is not yet defined.

Mte90 replied:
fixed

toolstack commented:
Would changing the format of the "entries" from "test by user @ time" to:

> user @ time added:
> text
> &nbsp;

(including the blank line at the end) be more readable?

Likewise I think splitting the button in two makes it easier to understand.  One button under the old text when you are an admin to update that and the second button under the new text dedicated to adding new.

With the current structure how do I know what the update button does?  If I edit both fields is one going to overwrite the other?

Mte90 commented:
The code check what text is changed and automatically update but I changed the text of feedback as you suggested :-)

Mte90 commented:
For me a single button is more useful because I can maybe change the old notes to remove some of them and add a new one but when I click the button I got a refresh (ajax call that replace all the html of that section) so I can lost all the edits. With a single button the user has the suggestions that can change both of them with a single action and avoid this problem.

toolstack commented:
I'm not tied to either way, I try and keep buttons doing only one thing as a general rule of thumb though.

The more I look at the code though I'm convinced there needs to be a `thing` for the table, we don't do direct SQL calls (except in very limited circumstances) as a rule and using an object for the table would keep everything consistant.

Take a look at the `gp-includes/things/` directory and the base class in `gp-includes/thing.php` for how to start.

Mte90 commented:
![screenshot_20170903_155633](https://user-images.githubusercontent.com/403283/30003631-b4675cc0-90c0-11e7-9ca8-ff01dcba5da5.png)
Two buttons and create a `things` :-)

Mte90 commented:
TODO: Replace feedback with note in all the code

Mte90 commented:
replaced the reference from feedback to note

Mte90 commented:
I fixed also the code issue in my code using the phpcs ruleset but phpcs report many other issue that are not from my code but in the files that I changed.

toolstack commented on gp-includes/things/note.php:
```diff
@@ -0,0 +1,95 @@
+ * @since 3.0.0
+ */
+class GP_Note extends GP_Thing {
+
```
You need to include the `$table_basename`, `$field_names`, `$int_fields` and `$non_updatable_attributes` properties for the `thing`, otherwise it will error out on first install.

Mte90 replied:
done :-)

toolstack commented on gp-includes/things/note.php:
```diff
@@ -0,0 +1,95 @@
+	 *
+	 * @return object The output of the query.
+	 */
+	function save_note( $note, $translation ) {
```
have `_note` is probably redundant.

toolstack commented on gp-includes/things/note.php:
```diff
@@ -0,0 +1,95 @@
+	 *
+	 * @return string The note.
+	 */
+	function get_note( $entry ) {
```
Same as above.

Mte90 committed:
little improvements as requested

Mte90 committed:
phpcs fixes

superdav42 commented:
@toolstack Any reason this wasn't merged? I'm looking to start a project that could benefit from these features.

Mte90 commented:
We are looking to release ASAP the 3.0 because this feature is scheduled for 3.1.
Maybe for the end of the month we can release the 3.0 and start to work finally on 3.1

superdav42 committed:
Normalize notes to use one row per note to allow more flixabilty is presentation

Each note row body only contains the contents of the note.
Which makes it possible to change the presentation without updating
all the rows is the database.
Each note can be deleted and edited individually.
Also displays the date in more human readible format.

akirk commented:
Am I reading this correctly that only people who can approve translations (i.e. Translation Editors) can comment? Does this mean that the original author of a translation doesn't have the ability to respond to a reviewer's comment?

Mte90 commented:
You are right this field is for who approve to leave comments about the translation. If the translator need more info can always reach the reviewer on slack (in the case of wp.org).

We want to avoid that this field became a chat also because there are no notifications at the moment about this comments.

garretthyder commented:
Hey @Mte90 just a thought from the peanut gallery.

With this screen you shared;
https://user-images.githubusercontent.com/403283/30003631-b4675cc0-90c0-11e7-9ca8-ff01dcba5da5.png

It appears you'll be able to edit other users comments to appear like they said something different. Might be better to only allow delete of other users comments so users can't change what someone else said.

But otherwise I can't wait to see this, thank you

superdav42 commented:
![screenshot from 2018-05-17 21-59-00](https://user-images.githubusercontent.com/1534605/40215443-c3b68f44-5a1d-11e8-9e22-baed6dc0c711.png)
This PR has been updated so notes are created, edited and deleted individually. The presentation is also improved.

toolstack commented:
I like the individual comments, but wonder if it may become too long within the UI without some kind of scroll bar for them?

Mte90 commented:
We want to plan for 3.0? I can update asap :-)
Also if we merge that we need to update https://github.com/GlotPress/GlotPress-WP/wiki/1.6-Roadmap
Maybe we can discuss about the roadmap

toolstack commented:
@Mte90 Yes, please update when you get a chance, I think this makes sense for 3.0.

Mte90 commented:
Talking about the comments I think that I can try to do a container with aon overflow so if there are a lot we can get a scrollbar.

I am evaluating also how to do the unit tests but I will check later :-)

Mte90 committed:
resolve conflicts

garretthyder commented on gp-templates/helper-functions.php:
```diff
@@ -315,24 +386,162 @@ function references( $project, $entry ) {
+ * @param string $translation_set The current translation set.
+ * @param string $location        The location of this toolbar, used to make id's unique for each instance on a page.
+ */
+function gp_translations_bulk_actions_toolbar( $bulk_action, $can_write, $translation_set, $location = 'top' ) {
```
This gp_translations_bulk_actions_toolbar function has been duplicated twice so there's currently 3 instances which causes a fatal;
[19-Oct-2018 05:13:06 UTC] PHP Fatal error:  Cannot redeclare gp_translations_bulk_actions_toolbar() (previously declared in /app/public/wp-content/plugins/glotpress/gp-templates/helper-functions.php:420) in /app/public/wp-content/plugins/glotpress/gp-templates/helper-functions.php on line 546

Mte90 replied:
my fault on resolving last conflicts

garretthyder commented on gp-includes/things/note.php:
```diff
@@ -0,0 +1,125 @@
+	 *
+	 * @return object The output of the query.
+	 */
+	function save( $note, $translation ) {
```
The GP_Thing save signature is public with a single $args parameter;
```suggestion
	public function save( $args = null ) {
```
The $note and $translation can live as args and carry through to the create function after you add the other items.

garretthyder commented on gp-includes/things/note.php:
```diff
@@ -0,0 +1,125 @@
+	 *
+	 * @return array notes
+	 */
+	function get_by_entry( $entry, $order = null ) {
```
Should be public function?
```suggestion
	public function get_by_entry( $entry, $order = null ) {
```

garretthyder commented:
Thinking on the overflow and potential for an excessive list. For longstanding projects where the history is useful but can become cumbersome an archive option can allow them to keep but hide old notes.

Commit:
another fix

Co-Authored-By: Mte90 <mte90net@gmail.com>

Commit:
another fix again

Co-Authored-By: Mte90 <mte90net@gmail.com>

Mte90 committed:
remove unused js var

Mte90 committed:
fix travis errors

Mte90 committed:
adding test calss for note route

Mte90 committed:
remove duplicate function

Mte90 committed:
first test

Mte90 committed:
fix linting but not yet finished

Mte90 committed:
fix linting but not yet finished

Mte90 committed:
fix linting but not yet finished

Mte90 committed:
fix linting but not yet finished

Mte90 committed:
fix linting and fix code

Mte90 committed:
fix linting and fix code

Mte90 committed:
fix linting and fix code

Mte90 committed:
adding scroll and fix linting code

Mte90 committed:
adding tests

Mte90 commented:
Let's see if the first tests are working :-D

Mte90 committed:
fixing for linting

Mte90 committed:
adding tests

Mte90 committed:
working on tests

Mte90 committed:
trying to find the class

Mte90 committed:
fix loading calss in unit tests

Mte90 committed:
fix loading class in unit tests

Mte90 committed:
finally some tests are working

toolstack commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,95 @@ function textareas( $entry, $permissions, $index = 0 ) {
+?>
+	<dl>
+		<dt>
+			<?php echo esc_attr__( 'Action Log:', 'glotpress' ); ?>
```
We should make this consistent in the UI, at the moment it's "Action Log" and "New Reviewer Note", perhaps "Notes" and "New note"?

Mte90 committed:
better text and fix alignments

toolstack commented:
Just taking a look at the visual layout (screenshots from Firefox):

![notes-meta](https://user-images.githubusercontent.com/9053749/47718250-5354dc00-dc1e-11e8-9929-c02ba0097f1b.png)

See how the "edit" link (which probably should be a button) gets cut off.

Likewise, if you shrink the screen width (to simulate a mobile device) the notes list doesn't use up the available space:

![notes-meta-2](https://user-images.githubusercontent.com/9053749/47718351-8f883c80-dc1e-11e8-8fc4-f311b49dc9b0.png)

And finally I'm wondering about the new note area, since editing is done inside of the list, could the new button be moved up above the list and the new area appear inside of it to cut down on some of the vertical height?

Last item is to discuss if "Add note" should move to the next translation or not, my gut so no but would love some feedback from actual translators :smile:

Mte90 commented:
My idea was to finish the tests and later work on the UI, so in that way the big part was already done and we can discuss freely about that.

toolstack commented:
Yep, just wanted to get my thoughts down now before I forgot them :wink:

Mte90 committed:
working on the other tests

Mte90 committed:
tests working!

Mte90 committed:
linting

Mte90 committed:
fix ui on mobile

Mte90 committed:
fix ui on mobile

ocean90 commented on gp-includes/routes/note.php:
```diff
@@ -0,0 +1,93 @@
+	 * @param object $translation Translation object.
+	 */
+	private function render_note( $note, $translation ) {
+		require_once GP_TMPL_PATH . 'helper-functions.php';
```
This should be avoided. The "back end" of GlotPress should not depend on functions provided by a theme.

Mte90 replied:
I am not sure how to do that because we are returning the html that is printed when someone use it the UI with ajax. 
We cannot force a refresh to see that change.

toolstack replied:
It should be done on the client side, the route to render a note shouldn't be required. 

You'll probably have to create a hidden template that you can copy when you need to render a note on the client side.

Mte90 replied:
done :-)

ocean90 commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,98 @@ function textareas( $entry, $permissions, $index = 0 ) {
+		</span>
+		<button class="note-actions" ><?php esc_attr_e( 'edit', 'glotpress' ); ?></button>
+		<div class="note-body">
+			<?php echo nl2br( esc_html( $note->note ) ); ?>
```
I'm assuming that many notes will include links so consider using `make_clickable()` here.

ocean90 commented on gp-includes/things/note.php:
```diff
@@ -0,0 +1,138 @@
+	 * @return object The output of the query.
+	 */
+	public function save( $args = null ) {
+		global $wpdb;
```
Unused import.

ocean90 commented on gp-includes/things/note.php:
```diff
@@ -0,0 +1,138 @@
+		global $wpdb;
+		$original_id    = gp_post( 'original_id' );
+		$translation_id = gp_post( 'translation_id' );
+		$note           = trim( gp_post( 'note' ) );
```
Before the note gets saved in the database it needs to be sanitized. I'd probably handling them the same as core does with comments.

ocean90 commented on gp-includes/things/note.php:
```diff
@@ -0,0 +1,138 @@
+		$translation_id = gp_post( 'translation_id' );
+		$note           = trim( gp_post( 'note' ) );
+		$translation    = new GP_Translation( array( 'id' => $translation_id ) );
+		if ( ! GP::$permission->current_user_can( 'approve', 'translation', $translation_id, array( 'translation' => $translation ) ) ) {
```
By default the "thing" classes should not include cap checks. This should be done before calling the save() method.

Mte90 replied:
Do you have any suggestion how to do that?
This is a class extended so I am not sure where I have to put that code.

toolstack replied:
You do it before you call the method, for example if your "saving" a note via a put route, then in the `new_post()` route method you'd do the permissions check before calling the save method on the note thing class.

ocean90 commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,98 @@ function textareas( $entry, $permissions, $index = 0 ) {
+?>
+	<div class="note">
+		<?php gp_link_user( get_userdata( $note->user_id ) ); ?>
+		<?php esc_attr_e( 'Commented', 'glotpress' ); ?>
```
`esc_attr_e()` should only be used for HTML attributes.

Mte90 replied:
the code style report that should be escaped

toolstack replied:
We don't have to escape translation functions, with the fix in #962 this should no longer throw a phpcs error.

ocean90 commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,98 @@ function textareas( $entry, $permissions, $index = 0 ) {
+ * @param GP_Translation $note          The note object.
+ * @param GP_Glossary    $can_edit      Permission of the user.
+ *
+ * @return number
```
`number` is not a valid PHP type.

ocean90 commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,98 @@ function textareas( $entry, $permissions, $index = 0 ) {
+			</button>
+			<button class="delete-note" tabindex="-1" data-note-id="<?php echo esc_attr( $note->id ); ?>" data-nonce="<?php echo esc_attr( wp_create_nonce( 'delete-note-' . $note->id ) ); ?>">
+				<?php esc_attr_e( 'Delete', 'glotpress' ); ?>
+			</button>
```
Maybe deleting a comment should only be possible by admins?

toolstack replied:
Or the original author.

I'm 50/50 that we shouldn't delete notes at all and just set a flag on them, but that seems like a lot of junk in the DB...

ocean90 commented:
IMO notes are not really content for the meta section. It's discussion which should have it's own section. Have you seen how WordPress.com is doing it?

<img width="1262" alt="comment-tab" src="https://user-images.githubusercontent.com/617637/47772894-7481ff00-dce8-11e8-8c33-944b0312e81e.png">

About the implementation:
* Why are notes linked to the original ID instead of the translation ID?
* How does someone know to which translation the feedback was for if there are multiple translations for one original?
* Should it be possible to let the translator reply to feedback of a validator?
* Should it be possible to let a translator add a note to their translation to provide additional context?
* Have you considered using WordPress' comments for feedback? This would give us edit/delete functions in the admin, spam protection through plugins like Akismet, integration into WordPress' GDPR tools, template functions for comments and the comment form, all for free.

cc: @yoavf @akirk IIRC you're using WordPress' comments for the discussions tab? Are the comments linked to the original or translation ID?

yoavf commented:
> IIRC you're using WordPress' comments for the discussions tab? Are the comments linked to the original or translation ID?

Yes, we're using WordPress comments. We're creating a shadow post in a custom post type that links to an original id. The reason we're doing this with an original id is that we thought that discussions could sometimes be effective across locales - i.e to discuss the original structure, or to get help and ideas for translating a complex term.

Mte90 committed:
fix typos and others stuff

Mte90 committed:
delete button only by capabilities

Mte90 commented:
* Why are notes linked to the original ID instead of the translation ID?

Probably a bug, in the code there are reference to use the translation id, if you saw something wrong let me know but I will do a check for it.

* How does someone know to which translation the feedback was for if there are multiple translations for one original?

As above

* Should it be possible to let the translator reply to feedback of a validator?

Good question that we never think of it, maybe for the next iteration we can think for this feature.

* Should it be possible to let a translator add a note to their translation to provide additional context?

I have to do a check because I am using the admin user right now (since I started to contributed I rebuilt my working instance)

* Have you considered using WordPress' comments for feedback? This would give us edit/delete functions in the admin, spam protection through plugins like Akismet, integration into WordPress' GDPR tools, template functions for comments and the comment form, all for free.

No, thinking of that this notes are for logged user, so they are already accepting the policy of the website and this space is only for the PTE and the author of localization.

Mte90 commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,100 @@ function textareas( $entry, $permissions, $index = 0 ) {
+		<div class="note-body">
+			<?php echo nl2br( make_clickable( esc_html( $note->note ) ) ); ?>
+		</div>
+		<?php if ( $can_edit || get_current_user_id() === $note->user_id ) : ?>
```
The check as the note user id is not useful, better add only PTE and the author of localization to avoid confusion.

Mte90 replied:
And do unit test for it with different permissions.

toolstack commented:
Speaking of cruft in the database, don't forget to update the deletion function of an original (or if you change it a translation) to also delete the associated notes.

ocean90 commented:
> Probably a bug, in the code there are reference to use the translation id, if you saw something wrong let me know but I will do a check for it.

In `GP_Note` search for `original_id`.

> No, thinking of that this notes are for logged user, so they are already accepting the policy of the website and this space is only for the PTE and the author of localization.

What policy? How is that related? Keep in mind that GlotPress has no "PTE", only validators and translators.

toolstack commented on gp-includes/router.php:
```diff
@@ -125,6 +125,10 @@ private function default_routes() {
+
+			'post:/notes/-new'                             => array( 'GP_Route_Note', 'new_post' ),
+			'post:/notes/-edit'                            => array( 'GP_Route_Note', 'edit_post' ),
+			'post:/notes/-delete'                          => array( 'GP_Route_Note', 'delete_post' ),
```
These lines are added in the middle of the set routes, move them further down to avoid any confusion.

toolstack commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,100 @@ function textareas( $entry, $permissions, $index = 0 ) {
+ *
+ * @return void
+ */
+function render_notes( $entry, $permissions ) {
```
Function name should be prepended with `gp_`.

toolstack commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,100 @@ function textareas( $entry, $permissions, $index = 0 ) {
+ * @param GP_Translation $entry            The translation entry.
+ * @param GP_Glossary    $permissions      Permissions of the user.
+ *
+ * @return void
```
When a function has no return value you don't need to add an @return line.

toolstack replied:
Looking at `render_note()`, it has a return value, these two functions should be consistent one way or the other.

toolstack commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,100 @@ function textareas( $entry, $permissions, $index = 0 ) {
+ *
+ * @return int
+ */
+function render_note( $note, $can_edit ) {
```
Function name should be prepended with `gp_`.

toolstack commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,100 @@ function textareas( $entry, $permissions, $index = 0 ) {
+ * @param GP_Translation $note          The note object.
+ * @param GP_Glossary    $can_edit      Permission of the user.
+ *
+ * @return int
```
This function returns an array, not an int.

toolstack commented on gp-templates/helper-functions.php:
```diff
@@ -287,6 +287,100 @@ function textareas( $entry, $permissions, $index = 0 ) {
+	return array(
+		'id'   => $note->id,
+		'note' => $note->note,
+	);
```
Is there a reason to return a value here?

Mte90 replied:
for unit tests to avoid the check the content of the html

Mte90 committed:
fix suggestions

Mte90 committed:
moving to templates

Mte90 committed:
moved to templates

Mte90 committed:
linting

Mte90 committed:
updated unit tests

Mte90 committed:
update for consistency

Mte90 committed:
silent error

Mte90 committed:
save translation id and use as reference

Mte90 committed:
permission check for translation author and validator

Mte90 commented:
I changed the data saved as `translation_id` instead `original_id` that was creating issues as described and also improved the permission check. 
Now check if is the author of translation or with with the capabilities for the notes action.
In that way is limited who can write to them.
I done also the other changes asked.

Missing only to delete the note referenced when the original is removed. If someone has suggestion where look for that I will do also that.
I think that I can use the hook `gp_original_deleted` but I am not sure if it is the right way or if it is better to change `original.php` on `after_delete`.

Mte90 committed:
fix linting

toolstack committed:
merge

toolstack commented:
> Missing only to delete the note referenced when the original is removed. If someone has suggestion where look for that I will do also that.
I think that I can use the hook gp_original_deleted but I am not sure if it is the right way or if it is better to change original.php on after_delete.

No need to use the hook, just add the code to after_delete would be my suggestion.

I also agree with @ocean90 that the notes should be moved out of the meta section, I've created a branch called `209-notes` with a layout that I think makes sense:

![3-way-layout](https://user-images.githubusercontent.com/9053749/48204159-2a77c980-e337-11e8-8e48-d532b7dd22aa.png)

There are a few other tweaks in the branch as well.

One thing I noticed is that currently saving notes is broken, the code inserts the note in to the database with a `null` value for the `translation_id`.

I also noticed that if you try and add a note to an original that does not have a translation, it fails silently.  There should be an error displayed in that case.

toolstack committed:
Update style for small screens

Mte90 committed:
remove note when original is removed

Mte90 committed:
remove unused line

Mte90 committed:
fix for unit tests

Mte90 committed:
show note editor if author or admin

Mte90 commented:
I fixed for original disabling the note editor, I am checking because is not saving the translation id but the ajax request is done.
So I am investigating and thanks for the new ui. I will upload it on my demo site :-)

Mte90 committed:
show note only on translation and not originals

Mte90 committed:
show notes only if there are

Mte90 committed:
improved unit tests

Mte90 committed:
missing translation_id now is saved

Mte90 committed:
merge from 209-notes

Mte90 committed:
fix linting

Mte90 committed:
rebase

Mte90 committed:
rebase

Mte90 committed:
rebase

Mte90 committed:
rebase

Mte90 committed:
rebase

Mte90 committed:
rebase

Mte90 committed:
rebase

Mte90 committed:
bump schema

Mte90 committed:
revert change

Mte90 committed:
revert change

Mte90 committed:
cleanup

La-Geek commented:
It is very appreciated that you are working on this problem 👍 
But this solution is very time consuming, PTEs/GTEs are faster, when they correct the error (the translator can compare his and the correctd string to find out, what was wrong).

This solution would be more effective:
https://github.com/GlotPress/GlotPress-WP/issues/324#issuecomment-384315517

Not to forget that there is the possibility of bulk edit. An example: The german language has two forms of address: default and formal. Often enough the strings are in wrong form of address ("Sie" in default e. g.). The workflow is: Filter for "Sie", tick all checkboxes at once and reject. 

Would be necessary to have also an textarea/reviewers feedback for bulk edit.

toolstack commented:
@La-Geek thanks for the feedback, the reject with feedback (aka the check boxes) will be looked at after 3.0, for now getting the notes/comments in place is a big step forward.

The current notes implementation isn't just for "reject with comments", but allows notes to be added at any time through the translation cycle.  This will allow translators to add information that could be useful in the future as well as to the current translation process.

toolstack commented:
Closing in favour of the clean git history version PR #975.
